### PR TITLE
Support source-removal logic in libbs

### LINKED ISSUE #103: Add BP_KEEP_FILES option for Java family buildpacks
## What happened?

* What were you attempting to do?
I have a Java server, which serves a front-end built by an embedded Yarn app (e.g., the `yarn` app lives in the `client` directory of the repo, and builds a `files/static/...` directory that the server provides to consumers). As such, I was trying to orchestrate a build by first running the `nodejs` buildpack, and then the `Java` buildpack. 

* What did you expect to happen?
I was expecting that the compiled front-end files would be available on the image, but because they were part of the source files (which are cleaned up by `libbs` [here](https://github.com/paketo-buildpacks/libbs/blob/96f9ae944c5244df2a2ec60c70f374f9b2495425/application.go#L136)), they were removed 

* What would you like?
I'd like to have an environment variable which functions similar to the `BP_KEEP_FILES` in Go, allowing users to specify source directories that `libbs` should not clean up

## Build Configuration

* What platform (`pack`, `kpack`, `tekton` buildpacks plugin, etc.) are you
  using? Please include a version.
I'm using `pack v0.23`

* What buildpacks are you using? Please include versions.
`paketo-buildpacks/java@5.21.0`
`paketo-buildpacks/nodejs@0.9.0`

* What builder are you using? If custom, can you provide the output from `pack
  inspect-builder <builder>`?
`paketobuildpacks/builder:full`

## Checklist

* [ ] I have included log output.
* [ ] The log output includes an error message.
* [ ] I have included steps for reproduction.

### PR DESCRIPTION
**Pull request by samj1912**
Signed-off-by: Sambhav Kothari <skothari44@bloomberg.net>

## Summary

Adds support for source-removal logic in libbs. Instead of completely empty-ing out the application directory, only relevant files are removed **BEFORE** copying back the interesting files to the application directory.

Fixes #103 

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Review by dmikusa — APPROVED** [review]

**Comment by dmikusa** [comment]
Looks good to me. This is a fairly core change though, so I'm going to let @pivotal-david-osullivan take a look as well. If you think it's good @pivotal-david-osullivan feel free to merge.

File: application.go [hunk 1]
```diff
@@ -133,17 +134,32 @@ func (a Application) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
 
 	// Purge Workspace
 	a.Logger.Header("Removing source code")
-	cs, err := ioutil.ReadDir(a.ApplicationPath)
-	if err != nil {
-		return libcnb.Layer{}, fmt.Errorf("unable to list children of %s\n%w", a.ApplicationPath, err)
+	includeDirs, iset := a.ArtifactResolver.ConfigurationResolver.Resolve("BP_INCLUDE_FILES")
+	if includeDirs != "" {
+		if err := logic.Include(a.ApplicationPath, includeDirs); err != nil {
+			return libcnb.Layer{}, err
```
Comment (pivotal-david-osullivan) [inline application.go:140]:
```suggestion
			return libcnb.Layer{}, fmt.Errorf("unable to perform source-removal 'include' \n%w", err)
```
Just a minor suggestion - we tend to wrap all errors for easier tracing!

File: application.go [hunk 2]
```diff
@@ -133,17 +134,32 @@ func (a Application) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
...
+		}
 	}
-	for _, c := range cs {
-		file := filepath.Join(a.ApplicationPath, c.Name())
-		if err := os.RemoveAll(file); err != nil {
-			return libcnb.Layer{}, fmt.Errorf("unable to remove %s\n%w", file, err)
+	excludeDirs, eset := a.ArtifactResolver.ConfigurationResolver.Resolve("BP_EXCLUDE_FILES")
+	if excludeDirs != "" {
+		if err := logic.Exclude(a.ApplicationPath, excludeDirs); err != nil {
+			return libcnb.Layer{}, err
```
Comment (pivotal-david-osullivan) [inline application.go:146]:
```suggestion
			return libcnb.Layer{}, fmt.Errorf("unable to perform source-removal 'exclude' \n%w", err)
```

**Comment by samj1912** [comment]
@pivotal-david-osullivan should be good to go :)

**Review by pivotal-david-osullivan — APPROVED** [review]

**Pull request merged**